File: app/lib/ui/pantallas/lotes.dart
/// Lista de lotes, creación y escaneo de QR (RF-LOT-01, RF-LOT-04).
library;

import 'package:flutter/material.dart';
import 'package:mobile_scanner/mobile_scanner.dart';

import '../../datos/bd/base_datos.dart';
import '../../nucleo/modelo/etapas.dart';
import '../../servicios.dart';
import '../comun/widgets.dart';
import '../tema.dart';
import 'detalle_lote.dart';

class PantallaLotes extends StatelessWidget {
  const PantallaLotes({super.key});

  @override
  Widget build(BuildContext context) {
    final s = ProveedorServicios.de(context);

    return Scaffold(
      appBar: AppBar(
        title: const Text('Lotes de cacao'),
        actions: [
          IconButton(
            tooltip: 'Escanear QR',
            icon: const Icon(Icons.qr_code_scanner),
            onPressed: () => _escanear(context),
          ),
        ],
      ),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: () => _nuevoLote(context),
        icon: const Icon(Icons.add),
        label: const Text('Nuevo lote'),
      ),
      body: StreamBuilder<List<Lote>>(
        stream: s.lotes.observarLotes(),
        builder: (context, snapshot) {
          final lotes = snapshot.data ?? const [];
          if (lotes.isEmpty) {
            return EstadoVacio(
              icono: Icons.inventory_2_outlined,
              titulo: 'Todavía no hay lotes',
              explicacion:
                  'Un lote es el cacao que llega junto y se procesa junto. '
                  'Crea el primero cuando recibas las mazorcas.',
              accion: FilledButton.icon(
                icon: const Icon(Icons.add),
                label: const Text('Crear el primer lote'),
                onPressed: () => _nuevoLote(context),
              ),
            );
          }
          return ListView.builder(
            padding: const EdgeInsets.fromLTRB(16, 8, 16, 96),
            itemCount: lotes.length,
            itemBuilder: (context, i) => _TarjetaLote(lote: lotes[i]),
          );
        },
      ),
    );
  }

  Future<void> _nuevoLote(BuildContext context) async {
    final s = ProveedorServicios.de(context);
    final fincas = await s.config.observarFincas().first;
    if (!context.mounted) return;

    String? fincaId = fincas.isEmpty ? null : fincas.first.id;
    DateTime fechaCosecha = DateTime.now();

    final crear = await showModalBottomSheet<bool>(
      context: context,
      isScrollControlled: true,
      builder: (contexto) => StatefulBuilder(
        builder: (contexto, actualizar) => Padding(
          padding: EdgeInsets.fromLTRB(
              16, 24, 16, MediaQuery.of(contexto).viewInsets.bottom + 24),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text('Nuevo lote',
                  style: Theme.of(contexto).textTheme.titleLarge),
              const SizedBox(height: 8),
              const Text(
                'El código se genera solo. Después registras las mazorcas.',
                style: TextStyle(fontSize: 15),
              ),
              const SizedBox(height: 20),
              if (fincas.isEmpty)
                const Aviso(
                  texto: 'No has registrado ninguna finca. Puedes crear el '
                      'lote igual y asignarle la finca después, desde Ajustes.',
                  icono: Icons.info_outline,
                  color: ColoresEstado.neutro,
                )
              else
                DropdownButtonFormField<String>(
                  initialValue: fincaId,
                  decoration: const InputDecoration(labelText: 'Finca'),
                  items: [
                    for (final f in fincas)
                      DropdownMenuItem(value: f.id, child: Text(f.nombre)),
                  ],
                  onChanged: (v) => actualizar(() => fincaId = v),
                ),
              const SizedBox(height: 16),
              ListTile(
                contentPadding: EdgeInsets.zero,
                leading: const Icon(Icons.event),
                title: const Text('Fecha de cosecha'),
                subtitle: Text(_fecha(fechaCosecha)),
                trailing: const Icon(Icons.edit),
                onTap: () async {
                  final elegida = await showDatePicker(
                    context: contexto,
                    initialDate: fechaCosecha,
                    firstDate: DateTime.now()
                        .subtract(const Duration(days: 60)),
                    lastDate: DateTime.now(),
                  );
                  if (elegida != null) {
                    actualizar(() => fechaCosecha = elegida);
                  }
                },
              ),
              const SizedBox(height: 20),
              BotonGrande(
                texto: 'Crear el lote',
                icono: Icons.check,
                onPressed: () => Navigator.pop(contexto, true),
              ),
            ],
          ),
        ),
      ),
    );

    if (crear != true || !context.mounted) return;
    final lote = await s.lotes.crearLote(
      fincaId: fincaId,
      fechaCosecha: fechaCosecha,
    );
    if (!context.mounted) return;
    avisar(context, 'Lote ${lote.codigo} creado');
    await Navigator.push(
      context,
      MaterialPageRoute(
          builder: (_) => PantallaDetalleLote(loteId: lote.id)),
    );
  }

  Future<void> _escanear(BuildContext context) async {
    final codigo = await Navigator.push<String>(
      context,
      MaterialPageRoute(builder: (_) => const PantallaEscaner()),
    );
    if (codigo == null || !context.mounted) return;

    final s = ProveedorServicios.de(context);
    // Un QR puede ser de lote (L-...), de saco (L-...-Sxx) o de tanda (P-...).
    final codigoLote = codigo.split('-S').first;
    final lote = await s.lotes.porCodigo(codigoLote);
    if (!context.mounted) return;

    if (lote == null) {
      avisar(context, 'No se encontró ningún lote con el código $codigo',
          error: true);
      return;
    }
    await Navigator.push(
      context,
      MaterialPageRoute(
          builder: (_) => PantallaDetalleLote(loteId: lote.id)),
    );
  }
}

class _TarjetaLote extends StatelessWidget {
  const _TarjetaLote({required this.lote});
  final Lote lote;

  @override
  Widget build(BuildContext context) {
    final s = ProveedorServicios.de(context);

    return Card(
      child: InkWell(
        borderRadius: BorderRadius.circular(16),
        onTap: () => Navigator.push(
          context,
          MaterialPageRoute(
              builder: (_) => PantallaDetalleLote(loteId: lote.id)),
        ),
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Row(
                children: [
                  Expanded(
                    child: Text(lote.codigo,
                        style: const TextStyle(
                            fontSize: 20, fontWeight: FontWeight.w700)),
                  ),
                  if (lote.ventaBloqueada)
                    const Chip(
                      avatar: Icon(Icons.block,
                          size: 18, color: ColoresEstado.problema),
                      label: Text('Venta bloqueada'),
                      visualDensity: VisualDensity.compact,
                    ),
                ],
              ),
              const SizedBox(height: 8),
              Row(
                children: [
                  Icon(Icons.timeline,
                      size: 18,
                      color: Theme.of(context).colorScheme.onSurfaceVariant),
                  const SizedBox(width: 6),
                  Text(lote.estado.etiqueta,
                      style: const TextStyle(
                          fontSize: 16, fontWeight: FontWeight.w600)),
                ],
              ),
              const SizedBox(height: 4),
              Text(lote.estado.queSigue,
                  style: TextStyle(
                      fontSize: 15,
                      color: Theme.of(context).colorScheme.onSurfaceVariant)),
              const SizedBox(height: 8),
              Row(
                children: [
                  Text('Llegó el ${_fecha(lote.fechaLlegada)}',
                      style: const TextStyle(fontSize: 14)),
                  const Spacer(),
                  StreamBuilder(
                    stream: s.alertas.observarAbiertas(loteId: lote.id),
                    builder: (context, snapshot) {
                      final n = snapshot.data?.length ?? 0;
                      if (n == 0) return const SizedBox.shrink();
                      return Chip(
                        avatar: const Icon(Icons.warning_amber_rounded,
                            size: 18, color: ColoresEstado.atencion),
                        label: Text('$n'),
                        visualDensity: VisualDensity.compact,
                      );
                    },
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}

/// Escáner de QR con guía visual (RF-LOT-04).
class PantallaEscaner extends StatelessWidget {
  const PantallaEscaner({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Escanear QR')),
      body: Stack(
        children: [
          MobileScanner(
            onDetect: (captura) {
              final codigo = captura.barcodes.firstOrNull?.rawValue;
              if (codigo != null && codigo.isNotEmpty) {
                Navigator.pop(context, codigo);
              }
            },
          ),
          Align(
            alignment: Alignment.bottomCenter,
            child: Container(
              width: double.infinity,
              color: Colors.black54,
              padding: const EdgeInsets.all(24),
              child: const Text(
                'Apunta al código QR del saco o de la barra.\n'
                'Se abrirá la historia completa del lote.',
                textAlign: TextAlign.center,
                style: TextStyle(color: Colors.white, fontSize: 16),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

String _fecha(DateTime f) =>
    '${f.day.toString().padLeft(2, '0')}/'
    '${f.month.toString().padLeft(2, '0')}/${f.year}';

File: android/app/src/main/kotlin/ec/cacaotrace/ui/pantallas/Lotes.kt
package ec.cacaotrace.ui.pantallas

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Block
import androidx.compose.material.icons.filled.Inventory2
import androidx.compose.material.icons.filled.QrCodeScanner
import androidx.compose.material.icons.filled.Timeline
import androidx.compose.material.icons.filled.WarningAmber
import androidx.compose.material3.AssistChip
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.ExtendedFloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavHostController
import ec.cacaotrace.ContenedorApp
import ec.cacaotrace.datos.bd.entidades.LoteEntidad
import ec.cacaotrace.nucleo.modelo.Codigos
import ec.cacaotrace.ui.ColoresEstado
import ec.cacaotrace.ui.Rutas
import ec.cacaotrace.ui.comun.BarraSuperior
import ec.cacaotrace.ui.comun.EstadoVacio
import ec.cacaotrace.ui.comun.Formato
import kotlinx.coroutines.launch

/** Lista de lotes, creación y escaneo de QR (RF-LOT-01, RF-LOT-04). */
@Composable
fun PantallaLotes(contenedor: ContenedorApp, navegacion: NavHostController) {
    val alcance = rememberCoroutineScope()
    val lotes by contenedor.lotes.observarLotes().collectAsState(initial = emptyList())
    var mensaje by remember { mutableStateOf<String?>(null) }

    // El escáner devuelve el código por el savedStateHandle de esta entrada.
    val entrada = navegacion.currentBackStackEntry
    val codigoEscaneado = entrada?.savedStateHandle
        ?.getStateFlow<String?>("codigoQr", null)?.collectAsState()

    LaunchedEffect(codigoEscaneado?.value) {
        val codigo = codigoEscaneado?.value ?: return@LaunchedEffect
        entrada?.savedStateHandle?.set("codigoQr", null)
        // Un QR puede ser de lote (L-…), de saco (L-…-Sxx) o de tanda (P-…).
        val lote = contenedor.lotes.porCodigo(Codigos.loteDeSaco(codigo))
        if (lote == null) {
            mensaje = "No se encontró ningún lote con el código $codigo"
        } else {
            navegacion.navigate(Rutas.detalleLote(lote.id))
        }
    }

    Scaffold(
        topBar = {
            BarraSuperior("Lotes de cacao") {
                IconButton(onClick = { navegacion.navigate(Rutas.ESCANER) }) {
                    Icon(Icons.Default.QrCodeScanner, contentDescription = "Escanear QR")
                }
            }
        },
        floatingActionButton = {
            ExtendedFloatingActionButton(
                onClick = {
                    alcance.launch {
                        val lote = contenedor.lotes.crearLote()
                        navegacion.navigate(Rutas.detalleLote(lote.id))
                    }
                },
                icon = { Icon(Icons.Default.Add, contentDescription = null) },
                text = { Text("Nuevo lote") },
            )
        },
    ) { relleno ->
        if (lotes.isEmpty()) {
            EstadoVacio(
                icono = Icons.Default.Inventory2,
                titulo = "Todavía no hay lotes",
                explicacion = "Un lote es el cacao que llega junto y se procesa junto. " +
                    "Crea el primero cuando recibas las mazorcas.",
                modifier = Modifier.fillMaxSize().padding(relleno),
                accion = {
                    Button(
                        onClick = {
                            alcance.launch {
                                val lote = contenedor.lotes.crearLote()
                                navegacion.navigate(Rutas.detalleLote(lote.id))
                            }
                        },
                    ) { Text("Crear el primer lote") }
                },
            )
            return@Scaffold
        }

        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(relleno),
            contentPadding = PaddingValues(16.dp, 8.dp, 16.dp, 96.dp),
        ) {
            mensaje?.let { texto ->
                item {
                    ec.cacaotrace.ui.comun.Aviso(
                        texto = texto,
                        color = ColoresEstado.problema,
                    )
                }
            }
            items(lotes, key = { it.id }) { lote ->
                TarjetaLote(contenedor, lote) {
                    navegacion.navigate(Rutas.detalleLote(lote.id))
                }
            }
        }
    }
}

@Composable
private fun TarjetaLote(
    contenedor: ContenedorApp,
    lote: LoteEntidad,
    alTocar: () -> Unit,
) {
    val alertas by contenedor.alertas.observarAbiertas(lote.id)
        .collectAsState(initial = emptyList())

    Card(modifier = Modifier.fillMaxWidth().padding(vertical = 6.dp).clickable(onClick = alTocar)) {
        Column(Modifier.padding(16.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    lote.codigo,
                    fontSize = 20.sp,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.weight(1f),
                )
                if (lote.ventaBloqueada) {
                    AssistChip(
                        onClick = alTocar,
                        leadingIcon = {
                            Icon(
                                Icons.Default.Block,
                                contentDescription = null,
                                tint = ColoresEstado.problema,
                            )
                        },
                        label = { Text("Venta bloqueada") },
                    )
                }
            }
            Spacer(Modifier.height(8.dp))
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    Icons.Default.Timeline,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Spacer(Modifier.height(6.dp))
                Text(
                    "  ${lote.estado.etiqueta}",
                    fontSize = 16.sp,
                    fontWeight = FontWeight.SemiBold,
                )
            }
            Text(
                lote.estado.queSigue,
                fontSize = 15.sp,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Spacer(Modifier.height(8.dp))
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    "Llegó el ${Formato.fecha(lote.fechaLlegada)}",
                    fontSize = 14.sp,
                    modifier = Modifier.weight(1f),
                )
                if (alertas.isNotEmpty()) {
                    AssistChip(
                        onClick = alTocar,
                        leadingIcon = {
                            Icon(
                                Icons.Default.WarningAmber,
                                contentDescription = null,
                                tint = ColoresEstado.atencion,
                            )
                        },
                        label = { Text("${alertas.size}") },
                    )
                }
            }
        }
    }
}
